clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -4687,6 +4687,10 @@ void CodeGenModule::HandleCXXStaticMemberVarInstantiation(VarDecl *VD) {
   if (DK == VarDecl::Definition && VD->hasAttr<DLLImportAttr>())
     return;
 
+  // Emit only those declarations that must be, when on SYCL device.
+  if (LangOpts.SYCLIsDevice && !MustBeEmitted(VD))
```

Comment:
This doesn't seem to be the right place for this.  This is the initialization, not when the variable itself is being emitted. 

Can you backtrace a little to see what is causing the static member to be emitted, and do the test there?  

Also, do we have the same problem with globals (non-members)?  How about magic-statics?

Reply:
What's happening with this is that the static variable emission (as well as globals and function-local statics) is being throttled using the usual mechanism in CodeGen and AST routines.  However, the initialization is being generated via a call from Sema:HandleCXXStaticMemberVarInstantiation() and the initialization later forces the variable to be emitted as well (since it gets used).

I have moved the check earlier in Sema.

Reply:
Can you point out the place where "initialization later forces the variable to be emitted as well"?

Reply:
HandleCXXStaticMemberVarInstantiation -> GetAddrOfGlobalVar -> getMangledName

And the code I mentioned above sees that getMangledName was previously called and emits the variable.

Reply:
What calls that handle-static-member-var-inistantiation?  We should probably be suppressing THAT unless necessary.

Reply:
That is called from Sema: SemaTemplateInstantiateDecl.cpp via PassToConsumeRAII but there isn't enough information there to suppress it.   I tried, but incorrectly; I couldn't figure out what property of Var I could use there.

Reply:
Perhaps we SHOULD be marking these vardecls as 'used-by-sycl' in some manner. If they aren't referenced in a kernel, we probably shouldn't be instantiating them, right?

That way we could just suppress the instantiate-variable-definition in PerformPendingInstantiations.

Reply:
Although, actually... If these are static-data-members, don't we KNOW that they can't be emitted?  We have an error that prevents them from being used, right?  So 'Is static member && SYCL' should be enough.  Either at the beginning of InstantiateVariableDefinition or in performpendinginstantiations, though I'm not sure where would be better.  See 6096 for how to check static-data-member.

That said, do Variable Templates cause us problems here too?

Reply:
We diagnose non-const static data members.  I will check out the other places you mentioned.

I haven't specifically tested variable templates, will do so.

Reply:
Ah, right.  Does this patch make sure that this gets emitted anyway if it is used then?  I'd hate for us to suppress this here, but have it be used later/before.